opsdroid/testing/__init__.py

```diff
@@ -0,0 +1,380 @@
+                            assert mock_api.called("/test")
+
+                    # Return True to we can ensure the closure was run
+                    return True
+
+                # Run the test closure. This will start the web server, and once it is running await
+                # the closure. Then when the closure finishes the server will be stopped again and the
+                # return from the closure will be passed back.
+                assert await mock_api.run_test(test)
+
+    """
+
+    def __init__(self):
+        """Initialize a server."""
+        self.app = None
+        self.runner = None
+        self._initialize_web_app()
+
+        self.site = None
+        self.host = "localhost"
+        self.port = 8089
+        self._calls = {}
+        self.responses = {}
+        self._payloads = {}
+        self.status = "stopped"
+
+    def _initialize_web_app(self) -> None:
+        self.app = web.Application()
+        self.runner = web.AppRunner(self.app)
+
+    async def _start(self) -> None:
+        """Start the server."""
+        await self.runner.setup()
+        self.site = web.TCPSite(self.runner, host=self.host, port=self.port)
+        await self.site.start()
+        self.status = "running"
+
+    async def _stop(self) -> None:
+        """Stop the web server."""
+        await self.runner.cleanup()
+        self.site = None
+        self.status = "stopped"
+
+    async def _handler(self, request: web.Request) -> web.Response:
+        route = request.path
+        if route in self._calls:
+            self._calls[route].append(request)
+        else:
+            self._calls[route] = [request]
+        if route in self._payloads:
+            self._payloads[route].append(await request.post())
+        else:
+            self._payloads[route] = [await request.post()]
+        status, response = self.responses[route].pop(0)
+        return web.json_response(response, status=status)
+
+    @property
+    def base_url(self) -> str:
+        """Return the base url of the web server."""
+        return f"http://{self.host}:{self.port}"
+
+    def add_response(
+        self, route: str, method: str, response_path: str, status: int = 200
+    ) -> None:
+        """Push a mocked response onto a route."""
+        if response_path is not None:
+            with open(response_path) as json_file:
+                response = json.load(json_file)
+        else:
+            response = None
+
+        if route in self.responses:
+            self.responses[route].append((status, response))
+        else:
+
+            if method.upper() == "GET":
+                routes = [web.get(route, self._handler)]
+            elif method.upper() == "POST":
+                routes = [web.post(route, self._handler)]
+            else:
+                raise TypeError(f"Unsupported method {method}")
+
+            self.responses[route] = [(status, response)]
+            self.app.add_routes(routes)
+
+    async def run_test(
```

Comment:
From a discussion in chat, this could be made a context manager which would lead to nicer tests imo.

Reply:
Agreed!